--test_arg broken in go_bazel_test

<!-- Please answer these questions before submitting your issue. Thanks! -->

### What version of rules_go are you using?

HEAD. This is in the main repo.
### What version of gazelle are you using?

The one in rules_go's WORKSPACE.

### What version of Bazel are you using?

0.29.1

If you try to pass a test argument to the go test inside of `go_bazel_test`, you'll get strange failures.

For instance, on that gopackagesdriver branch I've been committing on, I can run:

    bazel test //tests/core/gopackages:gopackages_test --test_output=all --test_arg='-v'

and the output will include this error message:

```
==================== Test output for //tests/core/gopackages:gopackages_test:
error: unexpected file: -v
panic: runtime error: invalid memory address or nil pointer dereference
================================================================================
```

and the target will have failed immediately.

Similar happens when I pass in `--test_arg='-test.v'` if the flagset stuff is important (the file name mentioned is, of course, `-test.v`).

--- a/go/tools/bazel_testing/bazel_testing.go
+++ b/go/tools/bazel_testing/bazel_testing.go
@@ -86,9 +86,32 @@ func TestMain(m *testing.M, args Args) {
 		os.Exit(code)
 	}()
 
+	var files []string
+	beginFiles, endFiles := -1, -1
+	for i, arg := range os.Args {
+		if arg == "-begin_files" {
+			beginFiles = i
+		} else if arg == "-end_files" {
+			endFiles = i
+			break
+		} else if arg == "--" {
+			break
+		}
+	}
+	if beginFiles >= 0 && endFiles < 0 ||
+		beginFiles < 0 && endFiles >= 0 ||
+		beginFiles >= 0 && beginFiles >= endFiles {
+		fmt.Fprintf(os.Stderr, "error: -begin_files, -end_files not set together or in order\n")
+		return
+	}
+	if beginFiles >= 0 {
+		files = os.Args[beginFiles+1 : endFiles-1]
+		os.Args = append(os.Args[:beginFiles], os.Args[endFiles+1:]...)
+	}
+
 	flag.Parse()
 
-	workspaceDir, cleanup, err := setupWorkspace(args)
+	workspaceDir, cleanup, err := setupWorkspace(args, files)
 	defer cleanup()
 	if err != nil {
 		fmt.Fprintf(os.Stderr, "error: %v\n", err)
@@ -177,7 +200,7 @@ func (e *StderrExitError) Error() string {
 	return sb.String()
 }
 
-func setupWorkspace(args Args) (dir string, cleanup func(), err error) {
+func setupWorkspace(args Args, files []string) (dir string, cleanup func(), err error) {
 	var cleanups []func()
 	cleanup = func() {
 		for i := len(cleanups) - 1; i >= 0; i-- {
@@ -244,7 +267,7 @@ func setupWorkspace(args Args) (dir string, cleanup func(), err error) {
 		runfileMap[runfileKey{rf.Workspace, rf.ShortPath}] = rf.Path
 	}
 	workspaceNames := make(map[string]bool)
-	for _, argPath := range flag.Args() {
+	for _, argPath := range files {
 		shortPath := path.Clean(argPath)
 		if !strings.HasPrefix(shortPath, "external/") {
 			return "", cleanup, fmt.Errorf("unexpected file: %s", argPath)

--- a/go/tools/bazel_testing/def.bzl
+++ b/go/tools/bazel_testing/def.bzl
@@ -35,7 +35,10 @@ def go_bazel_test(rule_files = None, **kwargs):
     # Add paths to rules_go files to arguments. bazel_testing will copy or link
     # these files.
     kwargs.setdefault("args", [])
-    kwargs["args"] += ["--"] + ["$(locations {})".format(t) for t in rule_files]
+    kwargs["args"] = (["-begin_files"] +
+                      ["$(locations {})".format(t) for t in rule_files] +
+                      ["-end_files"] +
+                      kwargs["args"])
 
     # Set rundir to the workspace root directory to ensure relative paths
     # are interpreted correctly.

